### ISSUE TITLE
Batch change for adding the Sourcegraph VS Code extension as recommended in VS Code configs

### ISSUE THREAD

**Issue by philipp-spiess** [sourcegraph/sourcegraph#40199]
- https://github.com/sourcegraph/batch-change-examples
- https://docs.sourcegraph.com/batch_changes/tutorials

/cc @muratsu @jjinnii @ryankscott

**Comment by philipp-spiess** [comment]
I started looking into this a bit and wanted to share my progress.

For the initial search, I think something like this should do the trick:

```
path:\.vscode
```

We could technically also search for `\.vscode\/extensions\.json$` but with the first version we can even create a config if none exists (ie if people only have other vscode specific settings saved).

Now for the actual manipulation I struggle a bit because Microsoft is using `jsonc` which is a different JSON spec that allows comments and trailing commas. This means that regular JSON clis like `json` won't be able to prase the files. I wasn't able to find any CLI that can structurally rewrite the jsonc files so far which means that we'll have to fallback to sed based text rewrite (anything else to parse the jsonc file will get rid of comments so the structure will be changed too much).

The text manipulation will need to be quite smart as well since it needs to add a new line if multiple lines are used to define recommendations and it needs to add the right indentation.

If a single line is used (e.g. `{"recommendations":[]}`) it may need to be able to break up the line into multiple lines.

I am now considering writing a custom script to do that since I ran out of other ideas 💺

**Comment by philipp-spiess** [comment]
Oooh I found the right library: https://github.com/microsoft/node-jsonc-parser

Of course we're already using it at Sourcegraph and of course we have contributed :D https://github.com/microsoft/node-jsonc-parser/graphs/contributors

This works:

```js
const fs = require("fs");
const { modify, applyEdits } = require("jsonc-parser");

const content = fs.readFileSync("extensions.json", "utf8").toString();

const edit = modify(
  content,
  ["recommendations", 0],
  "sourcegraph.sourcegraph",
  {
    isArrayInsertion: true,
    formattingOptions: {
      eol: "\n",
      insertSpaces: true,
      tabSize: 2,
    },
  }
);

console.log(applyEdits(content, edit));
```

**Comment by philipp-spiess** [comment]
This should do the trick, I am testing it now on https://github.com/sourcegraph/sourcegraph-search-export:

```
name: vscode-extension-recommendation-demo
description: Add a VS Code extension to the recommended extensions list

# Repositories containing this package
on:
  - repositoriesMatchingQuery: path:^\.vscode\/extensions.json$

steps:
  - run: |
      npm install -g jsonc-parser
      node /add-to-recommendations.js
    container: node:16-alpine
    files:
      /add-to-recommendations.js: |
        const fs = require("fs");
        const { modify, applyEdits, parse } = require("jsonc-parser");

        const EXTENSION_NAME = "sourcegraph.sourcegraph";

        const content = fs.readFileSync("extensions.json", "utf8").toString();
        const parsedContent = parse(content);

        const hasExtension =
          Array.isArray(parsedContent.recommendations) &&
          parsedContent.recommendations.includes(EXTENSION_NAME);

        if (hasExtension) {
          console.log("Extension already recommended");
          process.exit(0);
        }

        const edit = modify(
          content,
          ["recommendations", 0],
          "sourcegraph.sourcegraph",
          {
            isArrayInsertion: true,
            formattingOptions: {
              eol: "\n",
              insertSpaces: true,
              tabSize: 2,
            },
          }
        );

        fs.writeFileSync("extensions.json", applyEdits(content, edit));
        console.log("Extension added to recommendations");

changesetTemplate:
  title: Add recommended VS Code extension
  body: This extension provides a lot of value and we recommend it to all developers.
  branch: batch-changes/add-vscode-recommended-extension
  commit:
    message: Add recommended VS Code extension
  published: false
```

**Comment by philipp-spiess** [comment]
An error happened unfortunately but I’m going to continue next week. It looks like the module was not properly installed.

```
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.
added 1 package, and audited 2 packages in 418ms
found 0 vulnerabilities
npm notice 
npm notice New minor version of npm available! 8.11.0 -> 8.17.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.17.0>
npm notice Run `npm install -g npm@8.17.0` to update!
npm notice 
node:internal/modules/cjs/loader:936
  throw err;
  ^
Error: Cannot find module 'jsonc-parser'
Require stack:
- /add-to-recommendations.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/add-to-recommendations.js:2:39)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/add-to-recommendations.js' ]
}
Command failed with status 1
```

**Comment by philipp-spiess** [comment]
Updated the script, it works now. Will publish next week.
